File: src/main/java/net/dumbdogdiner/dogcore/commands/mute/MuteCommand.java
package net.dumbdogdiner.dogcore.commands.mute;

import dev.jorel.commandapi.annotations.Command;
import dev.jorel.commandapi.annotations.Default;
import dev.jorel.commandapi.annotations.Permission;
import dev.jorel.commandapi.annotations.arguments.AOfflinePlayerArgument;
import dev.jorel.commandapi.annotations.arguments.ATimeArgument;
import net.dumbdogdiner.dogcore.Permissions;
import net.dumbdogdiner.dogcore.database.User;
import net.dumbdogdiner.dogcore.messages.Messages;
import net.kyori.adventure.text.Component;
import org.bukkit.OfflinePlayer;
import org.bukkit.command.CommandSender;
import org.jetbrains.annotations.NotNull;

import java.time.Duration;
import org.jetbrains.annotations.Nullable;

@Command("mute")
@Permission(Permissions.MUTE)
public final class MuteCommand {
    /** Number of milliseconds per tick. */
    private static final long MILLISECONDS_PER_TICK = 50L;

    private MuteCommand() { }

    @Default
    public static void mute(
        final @NotNull CommandSender sender,
        final @NotNull @AOfflinePlayerArgument OfflinePlayer player,
        final @ATimeArgument int ticks
    ) {
        muteHelper(sender, player, Duration.ofMillis(ticks * MILLISECONDS_PER_TICK));
    }

    @Default
    public static void mute(
        final @NotNull CommandSender sender,
        final @NotNull @AOfflinePlayerArgument OfflinePlayer player
    ) {
        muteHelper(sender, player, null);
    }

    private static void muteHelper(
        final @NotNull CommandSender sender,
        final @NotNull OfflinePlayer player,
        final @Nullable Duration duration
    ) {
        User.lookupCommand(player, sender,
            user -> user.mute(duration).thenAccept(v -> user.formattedName().thenAccept(name -> {
                if (duration != null) {
                    var durationText = Component.text(duration.toString());
                    sender.sendMessage(Messages.get("commands.mute.duration", name, durationText));
                } else {
                    sender.sendMessage(Messages.get("commands.mute.indefinite", name));
                }
            })));
    }
}

File: src/main/kotlin/net/dumbdogdiner/dogcore/commands/MuteCommand.kt
package net.dumbdogdiner.dogcore.commands

import dev.jorel.commandapi.kotlindsl.anyExecutor
import dev.jorel.commandapi.kotlindsl.commandAPICommand
import dev.jorel.commandapi.kotlindsl.greedyStringArgument
import dev.jorel.commandapi.kotlindsl.offlinePlayerArgument
import net.dumbdogdiner.dogcore.Permissions
import net.dumbdogdiner.dogcore.db.DbPlayer
import net.kyori.adventure.text.Component
import org.bukkit.OfflinePlayer
import kotlin.time.Duration

fun muteCommand() = commandAPICommand("mute") {
    withPermission(Permissions.MUTE)

    offlinePlayerArgument("player")
    greedyStringArgument("duration", optional = true)

    anyExecutor { sender, args ->
        val player = DbPlayer.lookup(args["player"] as OfflinePlayer)
        if (player == null) {
            sender.sendMessage(Component.text("Player has no records in the server."))
            return@anyExecutor
        }

        val duration = args["duration"] as String?

        if (duration != null) {
            val parsedDuration = Duration.parseOrNull(duration)
            if (parsedDuration == null) {
                sender.sendMessage(Component.text("Failed to parse mute duration."))
            }
            player.mute(parsedDuration)
        } else {
            player.mute(null)
        }
    }
}
